File: src/main/java/no/nav/tjenestepensjon/simulering/config/TpLeverandorConfig.java
package no.nav.tjenestepensjon.simulering.config;

import static no.nav.tjenestepensjon.simulering.domain.TpLeverandor.EndpointImpl.valueOf;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import javax.validation.constraints.NotNull;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

import no.nav.tjenestepensjon.simulering.domain.TpLeverandor;

@Configuration
public class TpLeverandorConfig {

    private String leverandorUrlMap;

    @Value("${TP_LEVERANDOR_URL_MAP}")
    public void setLeverandorUrlMap(String leverandorUrlMap) {
        this.leverandorUrlMap = leverandorUrlMap;
    }

    @Bean
    public List<TpLeverandor> tpLeverandorList() {
        return createListFromEnv(leverandorUrlMap);
    }

    /**
     * Parse env variable to generate a list of TpLeverandor.
     * "," delimits the details of induvidual providers
     * "|" delimits different providers
     *
     * @param leverandorUrlMap env variable format "LEVERANDOR,URL,IMPL|..."
     * @return List of TpLeverandor
     */
    private List<TpLeverandor> createListFromEnv(@NotNull String leverandorUrlMap) {
        List<TpLeverandor> providers = new ArrayList<>();
        Arrays.stream(leverandorUrlMap.split("\\|")).forEach(provider -> providers.add(parseProvider(provider)));
        return providers;
    }

    private TpLeverandor parseProvider(@NotNull String provider) {
        String[] details = provider.split(",");
        assert details.length == 3;
        return new TpLeverandor(details[0], details[1], valueOf(details[2]));
    }
}

File: src/main/kotlin/no/nav/tjenestepensjon/simulering/config/TpLeverandorConfig.kt
package no.nav.tjenestepensjon.simulering.config

import no.nav.tjenestepensjon.simulering.model.domain.TpLeverandor
import org.slf4j.LoggerFactory
import org.springframework.boot.context.properties.ConfigurationProperties
import org.springframework.boot.context.properties.bind.Binder
import org.springframework.boot.context.properties.bind.ConstructorBinding
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Scope
import org.springframework.core.env.Environment

@Configuration
class TpLeverandorConfig(private val environment: Environment) {
    private val log = LoggerFactory.getLogger(javaClass)

    @Scope("singleton")
    @Bean("tpLeverandor")
    fun tpLeverandorList(): List<TpLeverandor> = PropertiesLoader(environment).let { propertiesLoader ->
        val list: MutableList<TpLeverandor> = mutableListOf()
        propertiesLoader.loadProperties("spk")?.let { list.add(it) }
        propertiesLoader.loadProperties("klp")?.let { list.add(it) }
        propertiesLoader.loadProperties("ofp")?.let { list.add(it) }
        propertiesLoader.loadProperties("gabler")?.let { list.add(it) }
        propertiesLoader.loadProperties("storebrand")?.let { list.add(it) }
        log.info("Loaded tp-leverandører: $list")
        return@let list.toList()
    }

    class PropertiesLoader(private val environment: Environment) {

        fun loadProperties(prefix: String): TpLeverandor? {
            return Binder.get(environment)
                .bind(prefix, TpLeverandorProperty::class.java)
                .map { TpLeverandor(it.name!!, it.implementation!!, it.simuleringUrl!!, it.stillingsprosentUrl!!) }
                .orElse(null)
        }
    }

    @ConfigurationProperties
    data class TpLeverandorProperty @ConstructorBinding constructor(
        val name: String?,
        val implementation: TpLeverandor.EndpointImpl?,
        val simuleringUrl: String?,
        val stillingsprosentUrl: String?
    )
}